### PR TITLE
contrib: Dump keys/certs from acme.json to files

### PR DESCRIPTION
**Pull request by brianredbeard**
In the event that a user needs to explode their acme.json file into
a set of directories and relevant files for troubleshooting or use
with other programs this script will parse them into the components
in the following path structure:

```
certdir
├── certs
│   ├── domain-1.example.com
│   ├── domain-2.example.com
│   └── domain-n.example.com
└── private
    └── letsencrypt.key
```

### CODE REVIEW

File: contrib/scripts/dumpcerts.sh [hunk 1]
```diff
@@ -0,0 +1,141 @@
+#!/bin/bash
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: `/usr/bin/env bash` is more portable and thus preferred.

File: contrib/scripts/dumpcerts.sh [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+# Exit codes:
+# 1 - A component is missing or could not be read
+# 2 - There was a problem reading acme.json
+# 4 - The destination certificate directory does not exist
+# 8 - Missing private key
+
+USAGE="dumpcerts.sh <path to acme> <destination cert directory>"
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: Can we use `$(basename "$0")` to display the script name in order to be portable should we ever rename the file?

File: contrib/scripts/dumpcerts.sh [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+# Exit codes:
+# 1 - A component is missing or could not be read
+# 2 - There was a problem reading acme.json
+# 4 - The destination certificate directory does not exist
+# 8 - Missing private key
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: Which exit codes one is allowed to use is a bit [disputed](http://tldp.org/LDP/abs/html/exitcodes.html); personally, I'm okay with using the "reserved" ones in the low-digit range.

Why powers of two though instead of just 1, 2, 3, and 4? We don't really need to shift numbers around, do we?

File: contrib/scripts/dumpcerts.sh [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+
+USAGE="dumpcerts.sh <path to acme> <destination cert directory>"
+
+# Allow us to exit on a missing jq binary
+exitJQ() {
+	echo ""
+	echo "You must have the binary 'jq' to use this."
+	echo "jq is available at: https://stedolan.github.io/jq/download/"
+	echo ""
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: Please output the message to stderr (`>&2`) using either a here document or a string stretching over multiple lines ([SO answer](http://stackoverflow.com/questions/10969953/how-to-output-a-multiline-string-in-bash)).

Same for other error outputs below.

File: contrib/scripts/dumpcerts.sh [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+# Exit codes:
+# 1 - A component is missing or could not be read
+# 2 - There was a problem reading acme.json
+# 4 - The destination certificate directory does not exist
+# 8 - Missing private key
+
+USAGE="dumpcerts.sh <path to acme> <destination cert directory>"
+
+# Allow us to exit on a missing jq binary
+exitJQ() {
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: I'd prefer a naming convention using only lower-case and underscores (`exit_jq`), like the [Google style guide recommends](https://google.github.io/styleguide/shell.xml?showone=Function_Names#Function_Names) as well.

File: contrib/scripts/dumpcerts.sh [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	echo ""
+	echo "You must have the binary 'jq' to use this."
+	echo "jq is available at: https://stedolan.github.io/jq/download/"
+	echo ""
+	echo ${USAGE}
+	exit 1
+}
+
+
+badACME() {
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: My preference goes towards `bad_acme` here too.

File: contrib/scripts/dumpcerts.sh [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+if [ ! -d "${certdir}" ]; then
+	echo ""
+	echo "Path ${certdir} does not seem to be a directory"
+	echo "We need a directory in which to explode the JSON bundle... exiting."
+	echo ""
+	echo ${USAGE}
+	exit 4
+fi	
+
+jq=$(which jq) || exitJQ
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: Please don't use `which` ([reasoning](http://stackoverflow.com/a/677212/420061)); use one of the alternatives.

File: contrib/scripts/dumpcerts.sh [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+badACME() {
+	echo ""
+	echo "There was a problem parsing your acme.json file."
+	echo ""
+	echo ${USAGE}
+	exit 2
+}
+
+
+acme="${1}"
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: Can we try to be a bit more specific? How about `acme_file`?

File: contrib/scripts/dumpcerts.sh (hunk 7 above)
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: Also, do we actually need to capture the result in a variable and use that one to invoke jq (as we do in line 85)? Why not invoke it directly once we know it's available?

File: contrib/scripts/dumpcerts.sh [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+# Exit codes:
+# 1 - A component is missing or could not be read
+# 2 - There was a problem reading acme.json
+# 4 - The destination certificate directory does not exist
+# 8 - Missing private key
+
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: Please set

```
set -o errexit
set -o pipefail
set -o nounset
```

to cover for common errors and have the script terminate prematurely.

File: contrib/scripts/dumpcerts.sh [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+	echo "Path ${certdir} does not seem to be a directory"
+	echo "We need a directory in which to explode the JSON bundle... exiting."
+	echo ""
+	echo ${USAGE}
+	exit 4
+fi	
+
+jq=$(which jq) || exitJQ
+
+priv=$(${jq} -r '.PrivateKey' ${acme}) || badACME
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: IMHO, we don't need the `|| badACME` part here and elsewhere below _if_ we enable `errexit`: If for some reason we cannot access the file, bash should bail out and give a meaningful error message.

File: contrib/scripts/dumpcerts.sh [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+mkdir -p "${certdir}"/{certs,private}
+
+pdir=$(realpath "${certdir}/private/")
+cdir=$(realpath "${certdir}/certs/")
+
+# Save the existing umask, change the default mode to 600, then
+# after writing the private key switch it back to the default
+oldumask=$(umask)
+umask 177
+# For some reason traefik stores the private key in stripped base64 format, but
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: It'd be nice to know the reason and phrase it here.

Maybe @emilevauge or @dtomcej know?

File: contrib/scripts/dumpcerts.sh (hunk 8 above)
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: I'd `readonly` the variable.

File: contrib/scripts/dumpcerts.sh [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+	echo ""
+	echo "There was a problem parsing your acme.json file."
+	echo ""
+	echo ${USAGE}
+	exit 2
+}
+
+
+acme="${1}"
+certdir="${2}"
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: I'd `readonly` the variable.

File: contrib/scripts/dumpcerts.sh [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+	echo "There didn't seem to be a private key in ${acme}."
+	echo "Please ensure that there is a key in this file and try again."
+	exit 8
+fi
+
+# If they do not exist, create the needed subdirectories for our assets
+# and place each in a variable for later use, normalizing the path
+mkdir -p "${certdir}"/{certs,private}
+
+pdir=$(realpath "${certdir}/private/")
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: Any particular reason we need to use `realpath`? AFAIK, it's not available on all systems.

File: contrib/scripts/dumpcerts.sh [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+cdir=$(realpath "${certdir}/certs/")
+
+# Save the existing umask, change the default mode to 600, then
+# after writing the private key switch it back to the default
+oldumask=$(umask)
+umask 177
```
Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: Shouldn't we use a `TRAP` to revert the umask in order to cover for the case where the script breaks anywhere between this line and the point where we return to the old umask?

**Comment by emilevauge** [comment]
@brianredbeard Wow great :+1: 
I think it would be even better if we could add a new command to Traefik (like `traefik exportACME /etc/ssl`). It would allow to export certs even in cluster mode (certs from the KV store). WDYT ?

File: contrib/scripts/dumpcerts.sh (hunk 3 above)
Comment (brianredbeard) [inline contrib/scripts/dumpcerts.sh]: This would just be so that in the long run you could continue through a number of failure situations and then give the user a composite at the end which could be used to identify all of the things that have gone wrong before doing anything (potentially) destructive.

**Comment by brianredbeard** [comment]
@emilevauge More than likely this would be a good idea, that being said this was a case of something that I was solving for myself anyways and figured could be useful for others.  I'd say making something like this available until a better answer is available (and then burning this script in a fire) is more than acceptable.

File: contrib/scripts/dumpcerts.sh (hunk 13 above)
Comment (brianredbeard) [inline contrib/scripts/dumpcerts.sh]: Was only really using this to normalize paths in the event of `${certdir}` being provided with a trailing slash.  I was of two minds about this.  `realpath` is a part of `coreutils` and thus should be on the lion's share of hosts (at least where traefik would be able to run). On the other hand it *is* a better answer to just guarantee this at the time of assignment via `${certdir%/}`... Admittedly this was me making this mostly for myself and thinking of it as "contrib" status work (YMMV).  Regardless, will fix with the assignment.

File: contrib/scripts/dumpcerts.sh (hunk 11 above)
Comment (emilevauge) [inline contrib/scripts/dumpcerts.sh]: @timoreimann what is your point exactly ? base64 ?

Comment (timoreimann) [inline contrib/scripts/dumpcerts.sh]: @emilevauge Yes. I think it would be nice to rephrase

_For some reason traefik stores the private key in stripped base64 format_

into something like

_traefik stores the private key in stripped base64 format because [reason]_

or just 

_traefik stores the private key in stripped base64 format_ (without explanation)

I don't think we should leave comments in scripts that express uncertainty about the reasoning in our own code if we can avoid it.

**Comment by ldez** [comment]
@brianredbeard It's clearly acceptable, can you fix the comments made by @timoreimann ? Thanks.

@emilevauge I've opened the issue #1512 for make the command in Traefik.

**Comment by emilevauge** [comment]
Ping @brianredbeard

**Comment by timoreimann** [comment]
@brianredbeard I carried your PR "in-place" and addressed all of my own comments that seemed undisputed. Hope that's okay with you. Feel free to comment back on me, of course.

@emilevauge PTAL.

**Comment by brianredbeard** [comment]
That's absolutely fine by me! Sorry about being a drag ass but you have my full support.

**Comment by ldez** [comment]
Due to SemaphoreCI, I close and reopen the PR.

**Comment by kachkaev** [comment]
Will this dump the certs automatically in real time? I'd like to mount a volume with my certs into gitlab and a mailserver, so am looking to find a folder with `pem` somewhere.

**Comment by frederik** [comment]
hi @brianredbeard. First of all, thanks for the contribution.

I have another (quick) follow up question. From how I understand let's encrypt it has two certificates. This script is extracting the domain handling key into a .key file and the certificate into the .pem file. But it doesn't extract the private key for the individual certificate (i.e. DomainsCertificate.Certs[n].Certificate.PrivateKey).

Gitlab Container Registry is asking me to provide both, so I did the base64 extraction manually to confirm that this was actually required.

I suppose extracting the PrivateKey would only be a minor duplication in the script.

**tl;dr**

Is there a reason you chose not to extract the PrivateKey? (or just didn't require it?)

**Comment by kachkaev** [comment]
@frederik could you please share your current workaround? I might need to pass a private key to gitlab or another upstream service in a near future :–)

**Comment by frederik** [comment]
I took the contents from DomainsCertificate.Certs[n].Certificate.PrivateKey and base64 decoded it by hand.

```
base64 -d <<< THECONTENTOF DomainsCertificate.Certs[n].Certificate.PrivateKey
-----BEGIN RSA PRIVATE KEY-----
...
-----END RSA PRIVATE KEY-----
```
(on linux -d, on Mac -D)

.. to a file and referenced the key in gitlab.

```
registry_nginx['ssl_certificate'] = "/etc/traefik/certs/my.domain.com.pem"
registry_nginx['ssl_certificate_key'] = "/etc/traefik/private/letsencrypt.key"
```

**Comment by ldez** [comment]
have you tried with the new version ? https://github.com/containous/traefik/pull/2161

**Comment by frederik** [comment]
I have now. This works for me. 👌

**Comment by poVoq** [comment]
Getting an error in line 71 with this
```
./dumpcerts.sh /opt/traefik/acme.json /opt/traefik/certs/
./dumpcerts.sh: Zeile 71: $1 ist nicht gesetzt.
```
"ist nicht gesetzt" means "is not set"

**Comment by torarnv** [comment]
Couldn’t this be a built in feature of traefik? Dumping the certs to a volume to be shared live with other software? Why require an external cron-driven export command to be issued, adding complexity for the use-case?

**Comment by ldez** [comment]
@torarnv https://github.com/containous/traefik/issues/1512

**Comment by torarnv** [comment]
@ldez Maybe I'm missing something, but #1512 still looks like a command that needs to be issued via `docker exec` in a running Træfik container, e.g. `traefik exportACME /etc/ssl`. Which means I need to drive it from a cron-job or similar manually, instead of just relying on Træfik to always keep a volume up to date with live certs. Isn't it just a Træfik built-in to replace the contrib-script?
